spark3-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/MergeIntoExec.scala

```diff
@@ -0,0 +1,111 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources.v2
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.{Attribute, BasePredicate, Expression, UnsafeProjection}
+import org.apache.spark.sql.catalyst.expressions.codegen.GeneratePredicate
+import org.apache.spark.sql.catalyst.plans.logical.MergeIntoParams
+import org.apache.spark.sql.execution.{SparkPlan, UnaryExecNode}
+
+case class MergeIntoExec(mergeIntoProcessor: MergeIntoParams,
+                         @transient targetRelation: DataSourceV2Relation,
+                         override val child: SparkPlan) extends UnaryExecNode {
+
+  override def output: Seq[Attribute] = targetRelation.output
+
+  protected override def doExecute(): RDD[InternalRow] = {
+    child.execute().mapPartitions {
+      processPartition(mergeIntoProcessor, _)
+    }
+  }
+
+  private def generateProjection(exprs: Seq[Expression], attrs: Seq[Attribute]): UnsafeProjection = {
+    UnsafeProjection.create(exprs, attrs)
+  }
+
+  private def generatePredicate(expr: Expression, attrs: Seq[Attribute]): BasePredicate = {
+    GeneratePredicate.generate(expr, attrs)
+  }
+
+  def applyProjection(predicates: Seq[BasePredicate],
+                      projections: Seq[UnsafeProjection],
+                      projectTargetCols: UnsafeProjection,
+                      projectDeleteRow: UnsafeProjection,
+                      inputRow: InternalRow,
+                      targetRowNotPresent: Boolean): InternalRow = {
+    // Find the first combination where the predicate evaluates to true
```

Comment:
I think this will require further explanation. There will be up to 2 matched cases and we try to find the first one that matches? Does `find` guarantee the order of the traversal?

Reply:
I think that `find` does guarantee the order and gives you the first one, which would be the first case that matches. I didn't see the syntax with multiple clauses and extra predicates in the SQL 2003 spec, so this might just be an extension that some implementations use. We should definitely check the semantics.

Reply:
@aokolnychyi @rdblue 
In my understanding, if we have overlapping conditions in two match branches i.e for example ..

WHEN MATCHED AND id > 0 and id < 10  UPDATE *
WHEN MATCHED AND id = 5  or id = 21 DELETE

Then we will just honor the first match and so if a row comes with id = 5, then we will update. And if a row with id 21 is present, then we will delete.